GPU tips-n-tricks webpage (#82)
mfem/web: src/gpu-tips-n-tricks.md

but more importantly the memory accesses on `A(i,j,k,n)` are much better with `MFEM_FORALL_3D`.
With `MFEM_FORALL_3D`, threads access consecutive memory, this is called coalesce memory access.
Because most applied math algorithms are highly memory bound, having coalesce memory accesses is critical to achieve high performance.

# Tips-n-tricks

## Compile in debug mode when developping for devices:
The memory manager performs checks that catches most of the missuses of the memory on host or device.
If using device debug, if your code fails you can run gdb or lldb, and set a breakpoint at `b mfem::mfem_error` .
The code will break as soon as it reaches this point and then you can backtrace `bt` from here to see what went wrong and where.

## Forcing synchronization with the host or the device:
It is sometimes needed to synchronize data between host and device.
In order to make sure that the host data is synchronized one should use `HostRead()`,
similarly to ensure synchronized data on the device one should use `Read()`.

## Do not use `GetData()`:

> Do not use `GetData()` to use a pointer on GPU since this will always return the host pointer wihtout synchronizing the data.

Reviewer: `wihtout` -> `without`

```c++
Vector v;
v.UseDevice(true);
const double *device_ptr = v.Read();
const double *host_ptr = v.HostRead();
```

## The MFEM_FORALL macro
The idea behind the `MFEM_FORALL` macro is to have the same behavior as a `for` loop and hide all the specifics to devices.
Example:
```c++
for(int i = 0; i < N; i++)
{
  ...
}
```
becomes
```c++
MFEM_FORALL(i, N,
{
  ...
});
```

The `DeviceTensor<N,T>` class is an `N` dimension array containing elements of type `T`, `T` is `double` by default.
It is convenient to use the `DeviceTensor<N,T>` class in combination with the memory manager and `MFEM_FORALL`.
The `Reshape` function reshapes an array into an `N` dimension array:
```c++
Vector a;
a.UseDevice(true);
const int p = ...;
const int q = ...;
const int r = ...;
const int N = ...;
auto A = Reshape(a.Write(), p, q, r, N); // returns a DeviceTensor<4,double>
MFEM_FORALL(n, N,
{
  for (int k = 0; k < r; k++)
    for (int j = 0; j < q; j++)
      for (int i = 0; i < p; i++)
        A(i,j,k,n) = ...;
});
```
There exists variants of this `MFEM_FORALL` macro, namely `MFEM_FORALL_2D` and `MFEM_FORALL_3D` which help maping 2D or 3D blocks of threads to the hardware more efficiently.

> In the case of a GPU, `MFEM_FORALL_3D(i,N,X,Y,Z,{...})` will declare `N` block of threads each of size `X`x`Y`x`Z` threads, whereas `MFEM_FORALL` uses `N` threads.

Reviewer 2: I don't believe the wording here of saying `N` threads are being used is correct for this section. If we look at the MFEM_FORALL case, MFEM by default only launches 256 threads to run the CUDA kernels.

Author: I don't think this is correct, `MFEM_FORALL` launches N/256 blocks of 256 threads, so N threads. I think it would be confusing to describe this, the fact that we use 256 threads per block is an implementation detail in this case I think.

Reviewer 2: Guess, it's been a while since I last looked at the actual documentation for how threading is done on CUDA devices, and I was wrong about my point earlier with the threads. However, you're still limited by the number of active concurrent threads that can be run at any one point: https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/index.html#differences-between-host-and-device . I know that number is hardware specific, so a max number shouldn't be put there. However, it might still be nice just to put a note in this section that all those threads may or may not be running concurrently depending on hardware limitations :)

Author: I see what you mean, but if I start talking about hardware threads, then I should probably talk about warps, SMs, etc... I agree that it is interesting and can be useful to understanding how to finely optimize code, but I think it would dive a bit too much into the hardware aspects of GPUs. I could add a note that by *thread* I always mean *software* thread, and never *hardware* thread.

Reviewer 2: I think a note about what you mean by thread would work for me.

Author: Thank you very much for you feedback, do not hesitate to point everything you find unclear/confusing/insufficient.

> # MFEM device handling

Reviewer 3: I would suggest opening the entire document with a short example, maybe something like this:
```c++
Vector u;
Vector v;
// ...
auto u_data = u.Read();
auto v_data = v.ReadWrite();
MFEM_FORALL({i, u.Size(),
{
   v_data[i] *= u_data[i];
});
```
Something like this IMO explains better the basics of the device support in MFEM better than the explanation; after a reader has seen this example, then the explanations of the memory manager and the forall macros will make more sense.

Author: Thank you, it's a good suggestion, I added comments to make it even more clear, tell me what you think.

MFEM relies mainly on two features for running algorithms on devices:

- The memory manager handles transparently moving data between host (CPU) and device (GPU for instance),

> - `MFEM_FORALL` allows to abstract `for` loops to parallelize transparently on an arbitrary device.

Reviewer 3: Someone told me that `MFEM_FORALL` is not about parallelism, it's about using a device...

Author: I meant MPI parallelism on that comment :)

## The memory manager
In order to make as transparent as possible the use of memory on host, or on device, MFEM relies on a memory manager.
Instead of storing a pointer of type `T*`, each object that can be accessed on a device contains `Memory<T>` objects.
The `Memory<T>` objects handle host and device pointers, memory allocations, and data synchronizations between host and device.

To get the pointer `T*` from a `Memory<T>` object, one has to use the `Read()`, `Write()`, or `ReadWrite()` methods.

-  `Read()` returns a `const T*` pointer, and should be used when the data will only be read,
-  `Write()` returns a `T*` pointer, and should be used when writing data **without** using any previously contained data,
-  `ReadWrite()` returns `T*` pointer, and should be used when read and write access to the pointer are required.

> `Read()`, `Write()`, and `ReadWrite()` are taking care automatically and completely transparently of data movements between host and device.

Reviewer 4: ```suggestion
`Read()`, `Write()`, and `ReadWrite()` automatically and transparently handles data movement between the host and device.
```

The method `void UseDevice(bool)` specifies if a `Memory<T>` object is intended for computation on host or on device.
The `Read()`,`Write()`, and `ReadWrite()` methods will return device pointer if using the device has been set to `true` with `UseDevice`, by default it is `false` and will return a host pointer.

Sometimes, it is necessary to access data on host regardless, in this situation the `HostRead()`, `HostWrite()`, and `HostReadWrite()` methods should be used.

In practice, developers rarely have to manipulate `Memory<T>` objects, instead objects data can be stored using `Vector` and `Array<T>`.
`Vector` and `Array<T>` data pointers can be accessed with the same methods as `Memory<T>`.
```c++
Vector v;
v.UseDevice(true);
const double *device_ptr = v.Read();
const double *host_ptr = v.HostRead();
```

## The MFEM_FORALL macro

> The idea behind the `MFEM_FORALL` macro is to have the same behavior as a `for` loop and hide all the specifics to devices.

Reviewer 4: ```suggestion
The idea behind the `MFEM_FORALL` macro is to have the same behavior as a `for` loop and hide all device-specific code in order to enable performance portability.
```

Example:
```c++
for(int i = 0; i < N; i++)
{
  ...
}
```
becomes
```c++
MFEM_FORALL(i, N,
{
  ...
});
```

The `DeviceTensor<N,T>` class is an `N` dimension array containing elements of type `T`, `T` is `double` by default.
It is convenient to use the `DeviceTensor<N,T>` class in combination with the memory manager and `MFEM_FORALL`.
The `Reshape` function reshapes an array into an `N` dimension array:
```c++
Vector a;
a.UseDevice(true);
const int p = ...;
const int q = ...;
const int r = ...;
const int N = ...;
auto A = Reshape(a.Write(), p, q, r, N); // returns a DeviceTensor<4,double>
MFEM_FORALL(n, N,
{
  for (int k = 0; k < r; k++)
    for (int j = 0; j < q; j++)
      for (int i = 0; i < p; i++)
        A(i,j,k,n) = ...;
});
```
There exists variants of this `MFEM_FORALL` macro, namely `MFEM_FORALL_2D` and `MFEM_FORALL_3D` which help mapping 2D or 3D blocks of threads to the hardware more efficiently.
In the case of a GPU, `MFEM_FORALL_3D(i,N,X,Y,Z,{...})` will declare `N` block of threads each of size `X`x`Y`x`Z` threads, whereas `MFEM_FORALL` uses `N/256` block of threads each of size `256` threads.

> Using `MFEM_FORALL_3D` (and `MFEM_FORALL_2D`) over `MFEM_FORALL` results in a higher level of parallelism, the former using `N`x`X`x`Y`x`Z` software threads and the later only `N` software threads.

Reviewer 5: later -> latter

In order to exploit 2D or 3D blocks of threads, it is convenient to use the macro `MFEM_FOREACH_THREAD(i,x,p)` to use threads as a `for` loop,
the first variable `i` is the name of the "loop" variable, `x` is the threadId, it can take the values `x`, `y`, or `z`, and `p` is the bound of the loop.
If we rewrite the previous example using `MFEM_FORALL_3D` and `MFEM_FOREACH_THREAD`, we get:
```c++
Vector a;
a.UseDevice(true);
const int p = ...;
const int q = ...;
const int r = ...;
const int N = ...;
auto A = Reshape(a.Write(), p, q, r, N); // returns a DeviceTensor<4,double>
MFEM_FORALL_3D(n, N, p, q, r,
{
  MFEM_FOREACH_THREAD(k,z,r)
    MFEM_FOREACH_THREAD(j,y,q)
      MFEM_FOREACH_THREAD(i,x,p)
        A(i,j,k,n) = ...;
});
```
The reasons for this more complex syntax is to better utilize the hardware, GPUs in particular.
Using `MFEM_FORALL_3D` and `MFEM_FOREACH_THREAD` allows to use more concurrency `NxXxYxZ` threads instead of only `N` threads with `MFEM_FORALL`,
but more importantly the memory accesses on `A(i,j,k,n)` are much better with `MFEM_FORALL_3D`.

> With `MFEM_FORALL_3D`, threads access consecutive memory, this is called coalesce memory access.

Reviewer 5: coalesce -> coalesced

> Because most applied math algorithms are highly memory bound, having coalesce memory accesses is critical to achieve high performance.

Reviewer 4: ```suggestion
Because most applied math algorithms are highly memory bound, having coalesced memory accesses is critical to achieve high performance.
```

Maximizing the main memory bandwidth is achieved by issuing enough memory transactions and using efficiently the data transferred.
The more computationally light a kernel is the more frequently memory transactions are issued, and if there is no memory bandwidth saturated other than the main memory bandwidth, e.g: shared or L1 memory, then the first condition to achieve peak performance is fulfilled.
Memory is transferred by contiguous blocks, called *cache-line*, which are typically the size of 32 `float`, or 16 `double`.
Since each cache-line is a block of contiguous memory it is common to over-fetch data when accessing non-contiguous memory addresses (because not all the data is used in each cache-line).
In the worst case, only one `float` of each cache-line is used resulting in only 1/32 of the data transferred being used, such a kernel is potentially 32 times slower than a kernel that would fully utilize the data in each cache line.
When a kernel is cautiously written to use all the data from each cache-line, the memory access are often referred as coalesce memory access.

> Having coalesces memory access kernels is critical to achieving peak performance.

Reviewer 5: coalesced

- Instruction Fetch — The next assembly instruction has not yet been fetched.
- Memory Throttle — A large number of pending memory operations prevent further forward progress. These can be reduced by combining several memory transactions into one.
- Memory Dependency — A load/store cannot be made because the required resources are not available or are fully utilized, or too many requests of a given type are outstanding. Memory dependency stalls can potentially be reduced by optimizing memory alignment and access patterns.
- Synchronization — The warp is blocked at a _syncthreads() call.
- Execution Dependency — An input required by the instruction is not yet available. Execution dependency stalls can potentially be reduced by increasing instruction-level parallelism.

You can use `nvprof --metrics` with:
`stall_inst_fetch` for the percentage of stalls occurring because of instruction fetch,
`stall_exec_dependency` for the percentage of stalls occurring because of execution dependency,
`stall_memory_dependency` for the percentage of stalls occurring because a memory dependency,

> `stall_memory_throttle` for the	percentage of stalls occurring because of memory throttle,

Reviewer 5: white space

## Roofline model
A [roofline model](https://developer.download.nvidia.com/video/gputechconf/gtc/2019/presentation/s9624-performance-analysis-of-gpu-accelerated-applications-using-the-roofline-model.pdf) helps predicting the peak performance achievable by a specific algorithm.
The arithmetic intensity is the ratio of the total number of operations divided by the amount of data movement from and to the main memory.
By dividing the maximum FLOPs, by the maximum bandwidth we get an arithmetic intensity threshold value between the two main regime of a GPU.
A kernel with an arithmetic intensity below and above the threshold value will be **memory bound** and **computation bound** respectively.

For in depths performance analysis I would recommend to look at [efficiency issues](https://docs.nvidia.com/gameworks/content/developertools/desktop/analysis/report/cudaexperiments/kernellevel/issueefficiency.htm)

[Here](https://docs.nvidia.com/cuda/profiler-users-guide/index.html#metrics-reference-7x) is the list of all the possible metrics for nvprof.

# Tips-n-tricks

## Compile in debug mode when developing for devices:

> The memory manager performs checks that catches most of the missuses of the memory on host or device.

Reviewer 5: misuse

When using device debug, if your code fails you can run gdb or lldb, and set a breakpoint at `b mfem::mfem_error`.
The code will break as soon as it reaches this point and then you can backtrace `bt` from here to see what went wrong and where.

## Forcing synchronization with the host or the device:
It is sometimes needed to force synchronization between host and device data.
In order to make sure that the host data is synchronized one should use `HostRead()`,
similarly to ensure synchronized data on the device one should use `Read()`.

## Do not use `GetData()`:
Do not use `GetData()` to access a pointer for device work since this will always return the host pointer without synchronizing the data with the device.

## Tracking data movements and allocations:

> Compiling MFEM with `MFEM_TRACK_CUDA_MEM` can help you see when data is transferred, allocated, etc.

Reviewer 4: It might help to mention how this helps you see the data transfers.

Ideally, "base" Vectors that will be modified/accessed on device through aliases should be allocated on device to begin with, e.g. using `Vector::SetSize(int s, MemoryType mt)` typically with `mt = Device::GetDeviceMemoryType()`.

## `MakeRef()` vectors do not see the same valid host/device data as their base vector:
```c++
const int vSize = 10;
Vector v; v.UseDevice(true); v.SetSize(vSize); v = 0.0;
cout << "IsHost(v) = " << IsHostMemory(v.GetMemory().GetMemoryType()) << endl;

Vector w; w.MakeRef(v, 0, vSize);
cout << "IsHost(w) = " << IsHostMemory(w.GetMemory().GetMemoryType()) << endl;

auto hv = v.HostWrite();
for (int j = 0; j < vSize; j++) { hv[j] = 1.0; }
cout << "IsHost(v) = " << IsHostMemory(v.GetMemory().GetMemoryType()) << endl;
cout << "IsHost(w) = " << IsHostMemory(w.GetMemory().GetMemoryType()) << endl;

Vector z; z.UseDevice(true); z.SetSize(vSize);
auto dz = z.Write();
auto dw = w.Read();
MFEM_FORALL(i, vSize,
{
  dz[i] = dw[i];
});
z.HostRead();
cout << "norm(z) = " << z.Norml2() << endl;

dz = z.Write();
auto dv = v.Read();
MFEM_FORALL(i, vSize,
{
  dz[i] = dv[i];
});
z.HostRead();
cout << "norm(z) = " << z.Norml2() << endl;

> ```

Reviewer 4: The output of the above example would add a lot of value to it, I think.

Author: I added the output, though I'm a bit confused by this example, @v-dobrev can you check that the output is what you expected, should we be observing some difference in the output?

Reviewer 6: Did you get this output with `"cuda"` device configuration? I'd expect different output in that case.
